Align protocol parameters JSON

v0d1ch opened the pull request:
> This is a reproduction of the issue we found where `cardano-cli query protocol-parameters` is missing to return certain fields (like `minFeeA`, `minFeeB`) which are present in the Json instance for `PParams` in `cardano-ledger`.  Since `cardano-cli` is just calling the `cardano-api` query to obtain `PParams` perhaps the bug is somewhere in the `fromLedgerPParams` or similar functions that convert back and forth between ledger and api representation of `PParams`.

@ch1bo / @locallycompact adopted this PR:

In fact, the `cardano-cli` was not broken, but the `hydra-node` was expecting `cardano-ledger` `PParams` on the `--ledger-protocol-parameters` input file .. which is wrong and this PR fixes this now. This was a _hidden_ breaking change on `master` so far!

* :space_invader: Load `--ledger-protocol-parameters` using the cardano-api `ProtocolParameters` instance, which is the one the `cardano-cli` also uses.

* :space_invader: Adds a test that asserts the `Hydra.Ledger.Cardano.Configuration.pparamsFromJson` function can read the JSON returned by `cardano-cli query protocol-parameters`

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [ ] No new TODOs introduced or explained herafter
  - One TODO about the fact that we should add (somewhere) and test against a JSON schema to ensure `pparamsFromJson` works as expected in `hydra-node` -> follow-up PR

v0d1ch committed:
Check cardano-cli version in scope and pparams roundtrip

v0d1ch committed:
Fix the failing test by setting the protocolVersion to optional field

This is needed until the upstram bug fix is released.

v0d1ch committed:
Query PParams using cardano-cli in scope and check compatibility

Move one related cardano-cli test to appropriate module

v0d1ch committed:
Contain the changes only in CardanoCliSpec

ch1bo committed:
Remove redundant cardano-cli test and add diagnostic output

The removed test is pointless as it does compare the output of using
cardano-api with using cardano-cli. We do not care whether those are
consistent, but rather that the output matches our expectations.

locallycompact committed:
CardanoCliSpec: Only test decoding to ProtocolParameters

ch1bo committed:
Draft a cliQueryProtocolParameters in withHydraNode'

This deliberately uses the cardano-cli to query protocol parameters as
JSON.

locallycompact committed:
match on ChainConfig in withHydraNode

ch1bo committed:
Fix invocation of cardano-cli and pin-point the root cause

ch1bo committed:
Add example protocol-parameters.json to hydra-cluster/config

These are a documented starting point to come up with parameters for the
hydra-node and are also used for the offline mode 'withHydraNode' in hydra-cluster.

locallycompact committed:
Remove PParams from withHydraNode calls

ch1bo committed:
Add blockTime to RunningNode to be flexible in e2e tests

locallycompact committed:
withHydraNode': Add Tracer argument

ch1bo committed:
Use 1s contestation period on full e2e test

This also improves the parse error message on hydra-node
--contestation-period as it helped us to debug this.

ch1bo committed:
Fix a typo

locallycompact committed:
withCardanoNode: callback with RunningNode instead of SocketPath

ch1bo committed:
Remove pparams from RunningNode

It's not a good idea to keep protocol parameters in the RunningNode
handle as they could change throughout our scenarios.

ch1bo committed:
Query genesis parameters to compute blockTime

locallycompact committed:
Infer blockTime properly

locallycompact committed:
Explicitly test pparamsFromJson in CardanoCliSpec

locallycompact committed:
simplify cardano-cli pparams test

ch1bo committed:
Upload benchmark logs

ch1bo committed:
Minor refactor while waiting for CI failures

ch1bo committed:
Debug benchmark failures in CI

ch1bo committed:
Use a dedicated work directory per dataset

ch1bo committed:
Add more trace output in bench

ch1bo committed:
Fix bench-e2e in CI by adding cardano-cli to shell

We were missing the cardano-cli from the shell environment running the
benchmark in CI.

The `withFile` from `base` is actively rewriting the ioe_location from
"posix_spawnp" to "withFile" and thus obfuscates any errors raised
within the wrapped action.

ffakenz commented on .github/workflows/ci-nix.yaml:
```diff
@@ -174,6 +174,14 @@ jobs:
+      uses: actions/upload-artifact@v4
+      with:
+        name: hydra-cluster-bench-logs
+        path: /tmp/nix-shell.*/bench-*/**/*.log
```
why is this needed?

ch1bo replied:
Used it for debugging to access log files when the bench fails (only cardano-node and hydra-nodes are not printed to github)

ffakenz commented on hydra-cluster/bench/Bench/EndToEnd.hs:
```diff
@@ -87,12 +87,13 @@ bench startingNodeId timeoutSeconds workDir dataset@Dataset{clientDatasets, titl
             hydraScriptsTxId <- seedNetwork node dataset (contramap FromFaucet tracer)
             let contestationPeriod = UnsafeContestationPeriod 10
-            withHydraCluster hydraTracer workDir nodeSocket startingNodeId cardanoKeys hydraKeys hydraScriptsTxId pparams contestationPeriod $ \(leader :| followers) -> do
+            putStrLn $ "Starting hydra cluster in " <> workDir
```
do we want to keep this print?

ch1bo replied:
Yeah it helped me debug and we have other prints around

ffakenz commented on hydra-cluster/bench/Bench/EndToEnd.hs:
```diff
@@ -238,14 +239,17 @@ seedNetwork :: RunningNode -> Dataset -> Tracer IO FaucetLog -> IO TxId
 seedNetwork node@RunningNode{nodeSocket, networkId} Dataset{fundingTransaction, clientDatasets} tracer = do
   fundClients
   forM_ clientDatasets fuelWith100Ada
+  putTextLn "Publishing hydra scripts"
```
maybe trace PublishedHydraScriptsAt instead?

ch1bo replied:
Don't want to do that now and it was mixing `putTextLn` and `traceWith` already.

ffakenz commented on hydra-cluster/bench/Bench/EndToEnd.hs:
```diff
@@ -238,14 +239,17 @@ seedNetwork :: RunningNode -> Dataset -> Tracer IO FaucetLog -> IO TxId
   publishHydraScriptsAs node Faucet
  where
   fundClients = do
+    putTextLn "Fund scenario from faucet"
```
maybe trace StartingFunds instead?

ffakenz commented on hydra-cluster/bench/Bench/EndToEnd.hs:
```diff
@@ -238,14 +239,17 @@ seedNetwork :: RunningNode -> Dataset -> Tracer IO FaucetLog -> IO TxId
 
   fuelWith100Ada ClientDataset{clientKeys = ClientKeys{signingKey}} = do
     let vk = getVerificationKey signingKey
+    putTextLn $ "Seed client " <> show vk
```
maybe create a trace for this?

ffakenz commented on hydra-cluster/bench/Main.hs:
```diff
@@ -10,69 +11,71 @@ import Bench.Options (Options (..), benchOptionsParser)
 
-  run outputDirectory timeoutSeconds startingNodeId targets = do
-    results <- forM targets $ \(dataset, dir) -> do
-      putStrLn $ "Test logs available in: " <> (dir </> "test.log")
```
why did we remove this print?

ch1bo replied:
Did not put it back after changing forth and back. I will put it back, moving it into the `bench` function.

ch1bo committed:
Minor refactor: Use readProcess and re-use cmd

ch1bo committed:
Not work around, but rely on withFile being fixed

Rebased the whole PR on top of
https://github.com/input-output-hk/hydra/pull/1236 and so we do not need
this workaround anymore.

ffakenz commented on hydra-cluster/src/CardanoClient.hs:
```diff
@@ -169,19 +167,20 @@ mkGenesisTx networkId pparams signingKey initialAmount recipients =
   | MsgCLIRetryResult Text Int
   | MsgNodeStarting {stateDirectory :: FilePath}
-  | MsgSocketIsReady FilePath
+  | MsgSocketIsReady SocketPath
```
why this change if `unFile SocketPath` returns FilePath?

is it because SocketPath displays better?

ch1bo replied:
I chose it because its a "stronger" type.

ffakenz commented on hydra-cluster/src/CardanoNode.hs:
```diff
@@ -283,20 +266,19 @@ withCardanoNode ::
   CardanoNodeArgs ->
   NetworkId ->
-  (SocketPath -> IO a) ->
+  (RunningNode -> IO a) ->
```
nice!

ffakenz commented on hydra-cluster/src/CardanoNode.hs:
```diff
@@ -308,12 +290,27 @@ withCardanoNode tr stateDirectory args@CardanoNodeArgs{nodeSocket} networkId act
     let nodeSocketPath = File socketPath
```
rename to nodeSocket?

ch1bo replied:
Considered it, but would be a conflict with the `CardanoNodeArgs` field. So I left it how it was.

ch1bo committed:
Print where benchmark logs are saved
